src/adapter/src/catalog/migrate.rs

```diff
@@ -182,6 +209,576 @@ pub(crate) async fn migrate(
+    // mat views, indexes) that reference a single-output source (e.g. kafka)
+    // will now reference the corresponding new table, with the same data, instead.
+    let mut changed_ids = BTreeMap::new();
+
+    for (mut item, stmt) in items_with_statements {
+        match stmt {
+            // Migrate each `CREATE SUBSOURCE` statement to an equivalent
+            // `CREATE TABLE .. FROM SOURCE` statement.
+            Statement::CreateSubsource(CreateSubsourceStatement {
+                name,
+                columns,
+                constraints,
+                of_source,
+                if_not_exists,
+                mut with_options,
+            }) => {
+                let raw_source_name = match of_source {
+                    // If `of_source` is None then this is a `progress` subsource which we
+                    // are not migrating as they are not currently relevant to the new table model.
+                    None => return Ok(()),
+                    Some(name) => name,
+                };
+                let source = match raw_source_name {
+                    // Some legacy subsources have named-only references to their `of_source`
+                    // so we ensure we always use an ID-based reference in the stored
+                    // `CREATE TABLE .. FROM SOURCE` statements.
+                    RawItemName::Name(name) => {
+                        // Convert the name reference to an ID reference.
+                        let (source_item, _) = items_with_statements_copied
+                            .iter()
+                            .find(|(_, statement)| match statement {
+                                Statement::CreateSource(stmt) => stmt.name == name,
+                                _ => false,
+                            })
+                            .expect("source must exist");
+                        RawItemName::Id(source_item.id.to_string(), name, None)
+                    }
+                    RawItemName::Id(..) => raw_source_name,
+                };
+
+                // The external reference is a `with_option` on subsource statements but is a
+                // separate field on table statements.
+                let mut i = 0;
+                let external_reference = loop {
+                    if i >= with_options.len() {
+                        panic!("subsource must have an external reference");
+                    }
+                    if with_options[i].name == CreateSubsourceOptionName::ExternalReference {
+                        let option = with_options.remove(i);
+                        match option.value {
+                            Some(WithOptionValue::UnresolvedItemName(name)) => break name,
+                            _ => unreachable!("external reference must be an unresolved item name"),
+                        };
+                    } else {
+                        i += 1;
+                    }
+                };
```

Comment:
Subjective, but something like the following might be slightly cleaner:

```Rust
    let option = match with_options
        .iter()
        .position(|opt| opt.name == CreateSubsourceOptionName::ExternalReference)
    {
        Some(i) => with_options.remove(i),
        None => panic!("subsource must have an external reference"),
    }

    let external_reference = match option.value {
        Some(WithOptionValue::UnresolvedItemName(name)) => name,
        _ => unreachable!("external reference must be an unresolved item name"),
    };
```

Reply:
Ah, TIL about `.position`!